concurrency/limited_concurrency_singleflight.go

```diff
@@ -0,0 +1,85 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+// LimitedConcurrencySingleFlight ensures that across all concurrent calls to ForEachNotInFlight, only up to maxConcurrent
+// tokens are running concurrently. See the docs of ForEachNotInFlight for the uniqueness semantics of tokens.
+type LimitedConcurrencySingleFlight struct {
+	inflightTokensMx sync.Mutex
+	inflightTokens   map[string]struct{}
+	inflightCalls    sync.WaitGroup
+	semaphore        chan struct{}
+}
+
+func NewWorkerPool(maxConcurrent int) *LimitedConcurrencySingleFlight {
+	return &LimitedConcurrencySingleFlight{
+		inflightTokensMx: sync.Mutex{},
+		inflightTokens:   make(map[string]struct{}),
+		inflightCalls:    sync.WaitGroup{},
+		semaphore:        make(chan struct{}, maxConcurrent),
+	}
+}
+
+// Wait returns when there are no in-flight calls to ForEachNotInFlight.
+func (w *LimitedConcurrencySingleFlight) Wait() {
+	w.inflightCalls.Wait()
+}
+
+// ForEachNotInFlight invokes f for every token in tokens that is not in-flight (not still being executed) in a different
```

Comment:
I'm assuming this is being added for a specific usecase in Mimir. Could we include some sort of example to make it clear when this would be useful?